### PR TITLE
[FLINK-25510][Connectors / Kafka][tests] Add using committed offsets test cases for KafkaPartitionSplitReader

### PR DESCRIPTION
**Pull request by ashulin**
## What is the purpose of the change

KafkaPartitionSplitReaderTest is missing the test case when startingOffset is COMMITTED_OFFSET.

## Brief change log

Add using committed offsets test cases for KafkaPartitionSplitReader

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java [hunk 1]
```diff
@@ -248,6 +253,100 @@ public void testAssignEmptySplit() throws Exception {
         assertTrue(recordsWithSplitIds.finishedSplits().isEmpty());
     }
 
+    @Test
+    public void testUsingCommittedOffsetsWithNoneOffsetResetStrategy() {
+        final Properties props = new Properties();
+        props.setProperty(
+                ConsumerConfig.GROUP_ID_CONFIG, "using-committed-offset-with-none-offset-reset");
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:260]: Maybe we can add a line of comment here, describing that you are using a new group ID without any committed offset, so an exception is expected

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java [hunk 2]
```diff
@@ -248,6 +253,100 @@ public void testAssignEmptySplit() throws Exception {
...
+        reader.fetch();
+        final Optional<Gauge<Long>> pendingRecords =
+                metricListener.getGauge(MetricNames.PENDING_RECORDS);
+        assertTrue(pendingRecords.isPresent());
+        // Validate pendingRecords
+        assertNotNull(pendingRecords);
+        assertEquals(NUM_RECORDS_PER_PARTITION - 1, (long) pendingRecords.get().getValue());
+        for (int i = 1; i < NUM_RECORDS_PER_PARTITION; i++) {
+            reader.fetch();
+            assertEquals(NUM_RECORDS_PER_PARTITION - i - 1, (long) pendingRecords.get().getValue());
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:314]: This reused code snippet actually is for testing the functionality of metric `pendingRecords`. I think a more straight-forward way under this case is to check if the first record that reader fetches is at the earliest offset, or check if the consuming position of the partition before any `fetch()` is the earliest offset.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java [hunk 3]
```diff
@@ -248,6 +253,100 @@ public void testAssignEmptySplit() throws Exception {
...
+        reader.handleSplitsChanges(
+                new SplitsAddition<>(
+                        Arrays.asList(latestOffsetResetEmptySplit, latestOffsetResetNormalSplit)));
+
+        // Fetch and check latest offset reset split is added to finished splits
+        RecordsWithSplitIds<ConsumerRecord<byte[], byte[]>> recordsWithSplitIds = reader.fetch();
+        assertTrue(
+                recordsWithSplitIds
+                        .finishedSplits()
+                        .contains(latestOffsetResetEmptySplit.splitId()));
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: I doubt that this case is testing the expected behavior. If the reader doesn't act as excepted, for example the fetcher doesn't respect the `auto.offset.reset` config and starts from the earliest offset, the first split (latestOffsetResetEmptySplit) could also reach the end offset in `reader.fetch()`, be added to finished splits, and pass this case. Also I can't see any purpose of the second split (latestOffsetResetNormalSplit) here.

Similar to the case above, I think a correct way is to check if the position before any `fetch()` is at the latest offset. This is the actual expected behavior of the reader.

**Review by PatrickRen — CHANGES_REQUESTED** [review]
Thanks for the patch @ashulin ! I left some comments.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java (hunk 3 above)
Comment (ashulin) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: `latestOffsetResetEmptySplit` will be added to the Finished split, but if there is no unfinished split `consumer.poll()` in the `reader.fetch()`  will throw an exception.
I also didn’t find out how to get the finished split without using `reader.fetch()` , so the test is done like this

Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: One possible solution: 
- Add a package-visible helper function to expose the consumer in `KafkaPartitionSplitReader`:
```java
@VisibleForTesting
KafkaConsumer<byte[], byte[]> consumer() {
    return consumer;
}
```
- Use `consumer.position()` in the test case to check the consuming offset before any fetch(), which is the actual starting offset of the reader.

Comment (ashulin) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: For the finished split, the reader will unassign the partition, and for the Kafka consumer, it is not allowed to get the offset of the unassign partition.

Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: What about making the split not bounded (not setting the ending offset)?

Comment (ashulin) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:347]: Thank you for your code review, I will fix soon.

**Review by PatrickRen — APPROVED** [review]
Thanks for the update @ashulin ! LGTM

@fapaul Could you help to merge the PR when you are available?

**Pull request merged**